### PR TITLE
[INLONG-4659][Sort] Support field routing for Elasticsearch connector

### LINKED ISSUE #4659: [Feature][Sort] Support routing field for Elasticsearch connector
### Description

Support routing field for Elasticsearch connector

### Use case

If routing.filed-name is set, will generate routing field for each IndexRequest, UpdateRequest and DeleteRequest.

### Are you willing to submit PR?

- [X] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by Oneal65**
### Prepare a Pull Request

- Fixes #4659 

### Motivation

Support field routing for ElasticSearch connector

### Modifications

according to  https://github.com/apache/flink/pull/14493/files
 and https://github.com/apache/flink/tree/release-1.13.2-rc2

modify the whole sort-connector-elasticsearch6 and sort-connector-elasticsearch7 code
and add sort-connector-elasticsearch-base module

### Background,  Design ideas  and others

Background:  the field routing concept, there are some URLs: 
https://www.cnblogs.com/bonelee/p/6055340.html
https://www.cnblogs.com/Rawls/p/10269192.html

Design ideas:

the routing function has been supported By ElasticSearch. What we should do is add this feature to the connector.

the main logic is in RoutingExtractor class

this PR is refer to:

https://github.com/apache/flink/pull/14493/files
https://github.com/apache/flink/tree/release-1.13.2-rc2

### Verifying this change

- [x] This change added tests and can be verified as follows:

  Elasticsearch6DynamicSinkITCase.java
  Elasticsearch7DynamicSinkITCase.java
  TestContext.java

### Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by GanfengTan** [comment]
1.Multiple module titles must correspond.
2.Commit history error.
image

File: .idea/vcs.xml [hunk 1]
```diff
@@ -1,36 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<!--
```
Comment (healchow) [inline .idea/vcs.xml:2]: Please do not change this file, it was used for quick links to GitHub from IDEA.

**Comment by dockerzhang** [comment]
@Oneal65 there are UTs failed, please check again.

**Comment by GanfengTan** [comment]
1.The problem background needs to be added.
2.Design ideas and schemes are necessary.
3.Add the copied source code to the LICENSE.

**Comment by Oneal65** [comment]
> 1.The problem background needs to be added. 2.Design ideas and schemes are necessary. 3.Add the copied source code to the LICENSE.



> 1.The problem background needs to be added. 2.Design ideas and schemes are necessary. 3.Add the copied source code to the LICENSE.



> 1.The problem background needs to be added. 2.Design ideas and schemes are necessary. 3.Add the copied source code to the LICENSE.

**Comment by Oneal65** [comment]
> @Oneal65 there are UTs failed, please check again.



> @Oneal65 there are UTs failed, please check again.



> @Oneal65 there are UTs failed, please check again.

File: .idea/vcs.xml (hunk 1 above)
Comment (Oneal65) [reply .idea/vcs.xml]: DONE

**Comment by Oneal65** [comment]
> 1.The problem background needs to be added. 2.Design ideas and schemes are necessary. 3.Add the copied source code to the LICENSE.

@GanfengTan Now, you can see it In Background, Design ideas and others

**Comment by Oneal65** [comment]
> @Oneal65 there are UTs failed, please check again.

@dockerzhang DONE

File: .idea/vcs.xml [hunk 2]
```diff
@@ -33,4 +33,4 @@
       </list>
     </option>
   </component>
-</project>
+</project>
```
Comment (GanfengTan) [inline .idea/vcs.xml:36]: Your local format is not accurate.

File: inlong-dashboard/package-lock.json [hunk 3]
```diff
@@ -11486,7 +11486,7 @@
     "json2mq": {
       "version": "0.2.0",
       "resolved": "https://registry.npmjs.org/json2mq/-/json2mq-0.2.0.tgz",
-      "integrity": "sha512-SzoRg7ux5DWTII9J2qkrZrqV1gt+rTaoufMxEzXbS26Uid0NwaJd123HcoB80TgubEppxxIGdNxCx50fEoEWQA==",
+      "integrity": "sha1-tje9O6nqvhIsg+lyBIOusQ0skEo=",
       "requires": {
         "string-convert": "^0.2.0"
       }
```
Comment (GanfengTan) [inline inlong-dashboard/package-lock.json:11492]: Only the functionality of a single module should be included under the Sort module.

File: inlong-sort/sort-connectors/elasticsearch-base/pom.xml [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+    <artifactId>sort-connector-elasticsearch-base</artifactId>
+    <name>Apache InLong - Sort-connector-elasticsearch-base</name>
+    <packaging>jar</packaging>
+
+
+    <properties>
+        <elasticsearch.version>6.8.17</elasticsearch.version>
+    </properties>
+
```
Comment (GanfengTan) [inline inlong-sort/sort-connectors/elasticsearch-base/pom.xml:38]: Put it in the project's dependency version management.

File: inlong-sort/sort-connectors/elasticsearch-base/src/main/java/org/apache/inlong/sort/elasticsearch/table/ElasticsearchOptions.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+                            "Delimiter for composite keys e.g., \"$\" would result in IDs \"KEY1$KEY2$KEY3\".");
+
+    public static final ConfigOption<String> ROUTING_FIELD_NAME =
+            ConfigOptions.key("routing.field-name")
+                    .stringType()
+                    .noDefaultValue()
+                    .withDescription("Elasticsearch routing filed.");
+
+    public static final ConfigOption<String> FAILURE_HANDLER_OPTION =
+            ConfigOptions.key("failure-handler")
```
Comment (GanfengTan) [inline inlong-sort/sort-connectors/elasticsearch-base/src/main/java/org/apache/inlong/sort/elasticsearch/table/ElasticsearchOptions.java:86]: Please add java doc for static method.

File: inlong-sort/sort-connectors/elasticsearch-base/src/main/java/org/apache/inlong/sort/elasticsearch/table/IndexGeneratorFactory.java [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+ * then a record with log_ts field value 2020-03-27 12:25:55 will be written into
+ * "myusers_2020-03-27" index.
+ */
+public final class IndexGeneratorFactory {
+
+    private IndexGeneratorFactory() {
+
+    }
+
+    public static IndexGenerator createIndexGenerator(String index, TableSchema schema) {
```
Comment (GanfengTan) [inline inlong-sort/sort-connectors/elasticsearch-base/src/main/java/org/apache/inlong/sort/elasticsearch/table/IndexGeneratorFactory.java:66]: Please add java doc

Comment (GanfengTan) [inline inlong-sort/sort-connectors/elasticsearch-base/src/main/java/org/apache/inlong/sort/elasticsearch/table/IndexGeneratorFactory.java:66]: Please add java doc

File: inlong-sort/sort-connectors/elasticsearch-base/src/main/java/org/apache/inlong/sort/elasticsearch/table/KeyExtractor.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+            return column.getType().getLogicalType();
+        }
+
+        public int getIndex() {
+            return index;
+        }
+    }
+
+    public static Function<RowData, String> createKeyExtractor(
+            TableSchema schema, String keyDelimiter) {
```
Comment (GanfengTan) [inline inlong-sort/sort-connectors/elasticsearch-base/src/main/java/org/apache/inlong/sort/elasticsearch/table/KeyExtractor.java:83]: Please add java doc.

**Review by GanfengTan — CHANGES_REQUESTED** [review]

File: .idea/vcs.xml (hunk 2 above)
Comment (Oneal65) [reply .idea/vcs.xml:36]: there is no changes in diff

File: inlong-dashboard/package-lock.json (hunk 3 above)
Comment (Oneal65) [reply inlong-dashboard/package-lock.json:11492]: It changed automatically.

File: inlong-sort/sort-connectors/elasticsearch-base/src/main/java/org/apache/inlong/sort/elasticsearch/table/ElasticsearchOptions.java (hunk 5 above)
Comment (Oneal65) [reply inlong-sort/sort-connectors/elasticsearch-base/src/main/java/org/apache/inlong/sort/elasticsearch/table/ElasticsearchOptions.java:86]: This is the Flink source code. we should not rewrite it for diff code in the future.

File: inlong-sort/sort-connectors/elasticsearch-base/src/main/java/org/apache/inlong/sort/elasticsearch/table/IndexGeneratorFactory.java (hunk 6 above)
Comment (Oneal65) [reply inlong-sort/sort-connectors/elasticsearch-base/src/main/java/org/apache/inlong/sort/elasticsearch/table/IndexGeneratorFactory.java:66]: the same reason.

Comment (Oneal65) [reply inlong-sort/sort-connectors/elasticsearch-base/src/main/java/org/apache/inlong/sort/elasticsearch/table/IndexGeneratorFactory.java:66]: the same reason.

File: inlong-sort/sort-connectors/elasticsearch-base/src/main/java/org/apache/inlong/sort/elasticsearch/table/KeyExtractor.java (hunk 7 above)
Comment (Oneal65) [reply inlong-sort/sort-connectors/elasticsearch-base/src/main/java/org/apache/inlong/sort/elasticsearch/table/KeyExtractor.java:83]: the same reason.

File: inlong-sort/sort-connectors/elasticsearch-base/pom.xml (hunk 4 above)
Comment (Oneal65) [reply inlong-sort/sort-connectors/elasticsearch-base/pom.xml:38]: DONE

File: inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/load/ElasticsearchLoadNode.java [hunk 8]
```diff
@@ -94,6 +97,7 @@ public ElasticsearchLoadNode(@JsonProperty("id") String id,
         this.index = Preconditions.checkNotNull(index, "index is null");
         this.documentType = documentType;
         this.primaryKey = primaryKey;
+        this.routingFieldName = primaryKey;
```
Comment (gong) [inline inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/load/ElasticsearchLoadNode.java:100]: `routingFieldName = primaryKey;`?

Comment (Oneal65) [reply inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/load/ElasticsearchLoadNode.java:100]: the default value for inlong-es-connector

Comment (gong) [reply inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/load/ElasticsearchLoadNode.java:100]: @Oneal65 User can not set it. Maybe `tableOptions` method can set defaut value for `routing.field-name`.

Comment (Oneal65) [reply inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/load/ElasticsearchLoadNode.java:100]: @gong OK

**Review by gong — APPROVED** [review]

**Review by thesumery — APPROVED** [review]

**Review by EMsnap — APPROVED** [review]

**Review by healchow — APPROVED** [review]

**Review by dockerzhang — APPROVED** [review]

**Pull request merged**